File: interview-prep/python-questions/src/binary-search-iterative.py
# Given a sorted array arr[] of n elements, write a function to search a given element x in arr[].
# Compare x with the middle element.
# If x matches with the middle element, we return the mid index.
# Else If x is greater than the mid element, then x can only lie in the right half subarray after the mid element. So we recur for the right half.
# Else (x is smaller) recur for the left half.

def binarySearchIterative(array, leftIndex, rightIndex, x):
    while rightIndex >= leftIndex:
        indexMid = leftIndex + (rightIndex - leftIndex) // 2 # floor division
        print(indexMid)
        if array[indexMid] == x:
            return indexMid
        elif array[indexMid] >= x:
            rightIndex = indexMid - 1
        else :
            leftIndex = indexMid + 1
    return -1


if __name__ == "__main__":
    arr = [2, 3, 4, 10, 40, 300]
    x = 101

    # Function call
    result = binarySearchIterative(arr, 0, len(arr) - 1, x)

    if result != -1:
        print("Element is present at index % d" % result)
    else:
        print("Element is not present in array")

File: interview-prep/java-questions/src/main/java/com/practice/code/searching/BinarySearchIterative.java
package com.practice.code.searching;

import com.practice.code.runner.CodeRunner;

public class BinarySearchIterative implements CodeRunner {

    // Returns index of x if it is present in arr[],
    // else return -1
    int binarySearch(int[] arr, int target) {
        int leftIndex = 0, rightIndex = arr.length - 1;
        while (leftIndex <= rightIndex) {
            int midIndex = leftIndex + (rightIndex - leftIndex) / 2;
            // Check if x is present at mid
            if (arr[midIndex] == target)
                return midIndex;
            // If x greater, ignore left half
            if (arr[midIndex] < target)
                leftIndex = midIndex + 1;
                // If x is smaller, ignore right half
            else
                rightIndex = midIndex - 1;
        }
        // if we reach here, then element was
        // not present
        return -1;
    }

    @Override
    public void run() {
        BinarySearchIterative ob = new BinarySearchIterative();
        int[] arr = {2, 3, 4, 10, 40};
        int n = arr.length;
        int x = 10;
        int result = ob.binarySearch(arr, x);
        if (result == -1)
            System.out.println("Element not present");
        else
            System.out.println("Element found at "
                    + "index " + result);
    }
}
